bash: Additional common suggestions (#2176)
exercism/website-copy: tracks/bash/exercises/raindrops/mentoring.md

## Common Suggestions

> * Check for a hash-pling: `#!/bin/bash`

Reviewer: This is more commonly known as a shebang (hash bang).

Author: # Bash Guidance

## Preflight checks

Check if the interpreter directive, aka hash-pling `#!` is present. Increased portability is acheived by not hard-coding this path. An indirect lookup with `env` is current best practice. Contrast:

```
#! /bin/bash
#! /usr/bin/env bash
```

## Interactive prompt self-help

Query if there an awareness of classic documentation that is typically available:

```
$ man  bash
$ info bash
```

A leaner alternative, [`tldr` (https://github.com/tldr-pages/tldr/)](https://github.com/tldr-pages/tldr/), reduces documentation down to common usage examples:

```
$ tldr bash

bash

Bourne-Again SHell, an `sh`-compatible command-line interpreter.
More information: <https://gnu.org/software/bash/>.

- Start an interactive shell session:
    bash

- Execute a command and then exit:
    bash -c "command"

...

```

## Static analysis & linting tools

[ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a stable pre 1.0 tool that provides a set of clear, editable rationales for each type of detected issue. Further less problematic recommendations can also be requested:

```
$ shellcheck <file>
$ shellcheck --enable=all <file>
```

## Debug

Consider Aaron Maxwell's [Bash Strict Mode (http://redsymbol.net/articles/unofficial-bash-strict-mode/)](http://redsymbol.net/articles/unofficial-bash-strict-mode/) to help reduce an invisible issue down to a minimal working example:

```
set -Eeuo pipefail
IFS=$'\n\t'
```

## Resources, Books and Printables

A special mention must be made of Michael Mol's chrestomathy site, [Rosetta Code (https://rosettacode.org/wiki/Category:Bash)](https://rosettacode.org/wiki/Category:Bash). Comparing solutions with other known or even unknown languages can be rather instructive.

|                            | Address
---------------------------- | ---------------------
Manual                       | https://www.gnu.org/software/bash/manual/
Frequently Asked Questions   | http://mywiki.wooledge.org/BashFAQ](http://mywiki.wooledge.org/BashFAQ
Reference Card               | https://devhints.io/bash
Pure BASH Bible              | https://github.com/dylanaraps/pure-bash-bible](https://github.com/dylanaraps/pure-bash-bible
Essential Bash               | https://www.programming-books.io/essential/
BASH Notes for Professionals | https://www.programming-books.io/essential/bash/
PDF version, 1.8 MB          | https://books.goalkicker.com/BashBook/BashNotesForProfessionals.pdf

* Consider a more robust hash-pling: `#!/usr/bin/env bash` 

> * Aaron Maxwell's [Bash Strict Mode](http://redsymbol.net/articles/unofficial-bash-strict-mode/): `set -euo pipefail`

Reviewer: The `set -e` is pretty controversial and has a load of pitfalls. Applying it blindly can be problematic.

Author: I removed it and mentioned it under Debug as a possible way to get a minimal working example.

        > Within an expression, shell variables may also be referenced by name without using the parameter expansion syntax.

      [https://www.gnu.org/software/bash/manual/bash.html#Shell-Arithmetic](https://www.gnu.org/software/bash/manual/bash.html#Shell-Arithmetic)

# Bash Guidance

## Preflight checks

> Check if the interpreter directive, aka hash-pling `#!` is present. Increased portability is achieved by not hard-coding this path. An indirect lookup with `env` is current best practice. Contrast:

Reviewer: ```suggestion
Check if the interpreter directive, aka hash-bash `#!` is present.
[Increased portability](http://mywiki.wooledge.org/BashGuide/CommandsAndArguments#Scripts) is achieved by not hard-coding this path, but looked up using `env`.
Contrast:
```

```
#! /bin/bash
#! /usr/bin/env bash
```

## Interactive prompt self-help

Query if there is an awareness of classic documentation that is typically available:

```
$ man  bash
$ info bash
```

> A leaner alternative, [`tldr` (https://github.com/tldr-pages/tldr/)](https://github.com/tldr-pages/tldr/), reduces documentation down to common usage examples:

Reviewer: `tldr` isn't a standard tool. As such, it shouldn't be a "common" suggestion.

```
$ tldr bash

bash

Bourne-Again SHell, an `sh`-compatible command-line interpreter.
More information: <https://gnu.org/software/bash/>.

- Start an interactive shell session:
    bash

- Execute a command and then exit:
    bash -c "command"

...

```

## Static analysis & linting

> [ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a stable pre 1.0 tool that provides a set of clear, editable rationales for each type of detected issue. Other less problematic recommendations can also be requested:

Reviewer: ```suggestion
[ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a tool that detects common issues
Other less problematic recommendations can also be requested:
```

```
$ shellcheck <file>
$ shellcheck --enable=all <file>
```

## Debug

Consider Aaron Maxwell's [Bash Strict Mode (http://redsymbol.net/articles/unofficial-bash-strict-mode/)](http://redsymbol.net/articles/unofficial-bash-strict-mode/) to help reduce an invisible issue down to a minimal working example:

```
set -Eeuo pipefail
IFS=$'\n\t'
```

## Resources

A special mention must be made of Michael Mol's chrestomathy site, [Rosetta Code (https://rosettacode.org/wiki/Category:Bash)](https://rosettacode.org/wiki/Category:Bash). Comparing solutions with other known or even unknown languages can be rather instructive.

> 

Reviewer: https://wiki.bash-hackers.org/scripting/tutoriallist might be worth a mention, as a curated/ranked list of tutorials.

Author: I'm adding it.

|                                  | Address
---------------------------------- | ---------------------
Manual                             | https://www.gnu.org/software/bash/manual/

> Frequently Asked Questions         | http://mywiki.wooledge.org/BashFAQ

Reviewer: ```suggestion
Wooledge Wiki         | http://mywiki.wooledge.org/
```
